coffea/lookup_tools/lookup_base.py

```diff
@@ -8,7 +10,33 @@ def __init__(self):
         pass
 
     def __call__(self, *args):
-        inputs = list(args)
+        if all(isinstance(x, (numpy.ndarray, numbers.Number)) for x in args):
+            return self._evaluate(*args)
+        elif any(isinstance(x, awkward.JaggedArray) for x in args):
+            return self._call_ak0(list(args))
+
+        def getfunction(inputs, depth):
+            if all(
+                isinstance(x, awkward1.layout.NumpyArray)
+                or not isinstance(
+                    x, (awkward1.layout.Content, awkward1.partition.PartitionedArray)
+                )
+                for x in inputs
+            ):
+                nplike = awkward1.nplike.of(*inputs)
+                if not isinstance(nplike, awkward1.nplike.Numpy):
+                    raise NotImplementedError("support for cupy/jax/etc. numpy extensions")
+                result = self._evaluate(*[nplike.asarray(x) for x in inputs])
+                return lambda: (awkward1.layout.NumpyArray(result),)
+            return None
+
+        behavior = awkward1._util.behaviorof(*args)
+        args = [awkward1.operations.convert.to_layout(arg, allow_record=False, allow_other=True) for arg in args]
+        out = awkward1._util.broadcast_and_apply(args, getfunction, behavior)
+        assert isinstance(out, tuple) and len(out) == 1
+        return awkward1._util.wrap(out[0], behavior)
```

Comment:
@jpivarski I lifted most of this from the numpy [ufunc](https://github.com/scikit-hep/awkward-1.0/blob/master/src/awkward1/_connect/_numpy.py#L40) implementation. Does it look general enough to put in awkward as a public method? I think maybe a nice extension would be some sort of single-dispatch based on `nplike`. In this context, we mandate all items are broadcastable numeric arrays, so I left out the `adjust` part, is that important in this context?

Reply:
Is `lookup_base` an abstract class and something else defines `_evaluate`?

Reply:
Correct, the derived classes implement `_evaluate(*numpy_arrays)`, e.g. https://github.com/CoffeaTeam/coffea/blob/master/coffea/lookup_tools/dense_lookup.py#L31